Sources/MapboxDirections/MatrixOptions.swift

```diff
@@ -0,0 +1,200 @@
+import Foundation
+import Turf
+
+/**
+ Options for calculating matrices from the Mapbox Matrix service.
+ */
+public class MatrixOptions: Codable {
+    // MARK: Creating a Matrix Options Object
+    /**
+     Initializes a matrix options object for matrices and a given profile identifier.
+     
+     - parameter waypoints: An array of `Waypoints` objects representing locations that will be in the matrix. The array should contain at least X coordinates and at most X coordinates. (Some profiles, such as `ProfileIdentifier`, [may have lower limits](https://docs.mapbox.com/api/navigation/matrix/#matrix-api-restrictions-and-limits).)
+     - parameter profileIdentifier: A string specifying the primary mode of transportation for the routes.
+     */
+    public init(waypoints: [Waypoint], profileIdentifier: ProfileIdentifier) {
+        self.waypoints = waypoints
+        self.profileIdentifier = profileIdentifier
+    }
+    
+    /**
+     A string specifying the primary mode of transportation for the contours.
+    */
+    public var profileIdentifier: ProfileIdentifier
+    
+    /**
+     An array of `Waypoints` objects representing locations that will be in the matrix.
+     */
+    public var waypoints: [Waypoint]
+    
+    /**
+     Attribute options for the matrix.
+
+     Empty `attributeOptions` will result in default values assumed.
+     */
+    public var attributeOptions: AttributeOptions = []
+    
+    /**
+     The coordinates at a given index in the `waypoints` array that should be used as destinations.
+     
+     By default, all waypoints in the `waypoints` array are used as destinations.
+     */
+    public var destinations: [Int]?
+    
+    /**
+     The coordinates at a given index in the `waypoints` array that should be used as sources.
+     
+     By default, all waypoints in the `waypoints` array are used as sources.
+     */
+    public var sources: [Int]?
```

Comment:
If we make a separate `Matrix.Waypoint` type, then it could have `isSource` and `isDestination` properties to avoid the need for maintaining parallel arrays of indices, which introduce the potential for out-of-bounds errors. Otherwise, a subclass of `Waypoint` might still be valuable as a place to put these properties.

Reply:
Re-worked this approach to have separate arrays of waypoints for sources and destinations.